java/arcs/core/data/expression/EvaluatorParticle.kt

```diff
@@ -90,39 +92,136 @@ class EvaluatorParticle(planParticle: Plan.Particle?) : BaseParticle() {
             val handle = handles.handles[name]
             val schema = requireNotNull(schemaMap[name])
             val expression = requireNotNull(metadata.handles[name]).expression!!
-            val result = evalExpression(expression, handlesScope)
+            val result = evalExpression(expression, handlesScope) ?: return
             when (hc.type) {
                 is SingletonType<*> -> {
                     @Suppress("UNCHECKED_CAST")
                     val singletonHandle = handle as WriteSingletonHandle<EntityBase>
-                    singletonHandle.store(entityFromScope(result as MapScope<*>, schema))
+                    singletonHandle.store(
+                        ScopeToEntityTranslator.translateEntity(schema, result)
+                    )
                 }
                 is CollectionType<*> -> {
                     @Suppress("UNCHECKED_CAST")
                     val collectionHandle = handle as WriteCollectionHandle<EntityBase>
                     @Suppress("UNCHECKED_CAST")
-                    (result as Sequence<MapScope<*>>).forEach {
-                        collectionHandle.store(entityFromScope(it, schema))
-                    }
+                    collectionHandle.storeAll(ScopeToEntityTranslator.translateCollection(
+                        // We piggy back on translation of sets of inline entities
+                        // to translate the expression result for a collection handle.
+                        CollectionOf(FieldType.InlineEntity(schema.hash)),
+                        result
+                    ) as Collection<EntityBase>)
                 }
             }
         }
     }
+}
+
+/** Extension of [FieldType] to represent a collection field. */
+private data class CollectionOf(val primitiveType: FieldType)
+
+/**
+ * Translator of entity data representations which can recurse down the nested structures
+ * while maintaining the mapping to the type description from the schema.
+ */
+private abstract class SchemaGuidedTranslator {
+
+    fun translateCollection(type: CollectionOf, value: Any): Set<Any> = when (value) {
+        is Set<*> -> value
+        is Sequence<*> -> value.toSet()
+        is Iterable<*> -> value.toSet()
+        else -> throw IllegalArgumentException("Unable to interpret $value as a Set")
+    }.map { translateSingleton(type.primitiveType, requireNotNull(it)) }.toSet()
+
+    fun translateSingleton(type: FieldType, value: Any): Any = when (type) {
+        is FieldType.ListOf -> translateList(type, value)
+        is FieldType.InlineEntity -> translateInlineEntity(type, value)
+        is FieldType.Primitive -> value
+        is FieldType.EntityRef -> TODO("References are not supported")
```

Comment:
Just a note, to avoid runBlocking in the future, these functions will need to be suspend functions since dereferencing will require it

Reply:
Interesting! Let's tackle it once we need to support references. We will need to either make Paxel lookups async or dereference everything eagerly. The is more efficient, but it's not a small refactoring.